### ISSUE TITLE
【问题】关于强化学习微调

### ISSUE THREAD

**Issue by freeclouds** [CurryChen77/RIFT#11]
hi 感谢开源了强化学习微调端到端网络的代码，我有一个问题就是我在分析目前开源的代码中，我理解是例如cbv使用的fift_pluto这个方式的微调，那么强化学习的作用是计算group_advantage；具体的流程是使用生成的ego和cbv轨迹通过action得到在当前环境的动作，然后调用carla接口self.env.step(ego_actions_dict, CBVs_actions_dict, ego_obs_list, CBVs_obs_list, info_list)执行这些action，通过运行这些action然后计算对应的group_advantage，再然后使用计算的group_advantage作为pluto轨迹概率网络的损失函数，损失函数即：
`     def get_rift_loss(self, probability, cur_data, batch):
        '''
        Compute the rift loss.
        '''
        bs, R, M = probability.shape  # (max_R*M=G)
        # only take the valid reference line probability
        cur_r_padding_mask = ~cur_data["reference_line"]["valid_mask"].any(-1)  # (bs, max_R)
        valid_mask = batch['group_advantage_mask_torch']  # (bs, max_R, M)
        r_padding_mask = ~valid_mask.any(-1)
        assert torch.equal(cur_r_padding_mask, r_padding_mask)

        # current probability and log probability
        probability.masked_fill_(cur_r_padding_mask.unsqueeze(-1), -1e8)
        log_action_prob = F.log_softmax(probability.view(bs, -1), dim=1)  # (bs, max_R * M)

        # Apply masks to old logits and compute log prob
        old_group_logits = batch['old_group_logits_torch']  # (bs, max_R, M)
        old_group_logits.masked_fill_(cur_r_padding_mask.unsqueeze(-1), -1e8)
        old_group_log_prob = F.log_softmax(old_group_logits.view(bs, -1), dim=1)  # (bs, max_R * M)
        
        # reshape group_advantage to (bs, G)
        group_advantage = batch['group_advantage_torch'].view(bs, -1)  # (bs, G)
        
        # Compute the truncated objective
        ratio = torch.exp(log_action_prob - old_group_log_prob)
        unclipped_objective = group_advantage * ratio
        clipped_objective = group_advantage * torch.clamp(ratio, 0.8, 1.2)
        min_objective = torch.min(unclipped_objective, clipped_objective)
        max_objective = torch.max(min_objective, group_advantage * 3.0)
        objective = torch.where(group_advantage < 0, max_objective, min_objective)

        # Compute surrogate loss considering only valid positions
        valid_mask_flat = valid_mask.view(bs, -1)  # (bs, G)
        valid_objective = objective[valid_mask_flat]
        if valid_objective.numel() == 0:
            surrogate = torch.tensor(0.0, device=probability.device)
        else:
            surrogate = valid_objective.mean()

        rift_loss = -surrogate

        return rift_loss`
其中损失函数里面的的group_advantage，然后将计算的group_advantage通过get_rift_loss损失来微调planning_decoder.pi_head网络(        self.pi_head = MLPLayer(dim, dim, 1))吗？也即grpo强化学习也就是如下代码：
`         
            CBV_episode_reward = []
            while not self.env.all_scenario_done():
                # get action from agent policy and scenario policy (assume using one batch)
                ego_actions_dict = self.ego_policy.get_action(ego_obs_list, info_list, deterministic=False)
                CBVs_actions_dict = self.cbv_policy.get_action(CBVs_obs_list, info_list, deterministic=False)
                # apply action to env and get obs
                data_dict = self.env.step(ego_actions_dict, CBVs_actions_dict, ego_obs_list, CBVs_obs_list, info_list)

                # store to the replay buffer
                buffer.store(data_dict)

                # for transition
                info_list = copy.deepcopy(data_dict['transition_info'])
                ego_obs_list = copy.deepcopy(data_dict['ego_transition_obs'])
                CBVs_obs_list = copy.deepcopy(data_dict['CBVs_transition_obs'])

                scenario_reward = np.concatenate([np.array(list(CBVs_reward.values())) for CBVs_reward in data_dict['CBVs_reward']])
                CBV_episode_reward.append(np.mean(scenario_reward)) if scenario_reward.size != 0 else None`
我不太理解强化学习的网络在整个训练代码里面是哪一部分，还是说强化学习只是计算pluto轨迹概率网络planning_decoder.pi_head的损失函数的一部分，没有对应的网络？

**Comment by CurryChen77** [comment]
我认为，在奖励函数的设计问题上，如果目标行为可以明确量化，那么应优先采用真实且准确的可量化奖励。例如 collision、off-road 等指标具有明确的判定标准和清晰的工程定义，使用此类奖励不仅能够降低设计复杂度，还能有效避免 reward hacking 等潜在问题。近期的研究也支持这一观点，[1] 指出在特定任务中，简单的奖励函数反而能带来更好的性能表现。

对于一些更难直接量化的驾驶行为，例如驾驶风格、驾驶意图交互等，奖励函数的设计通常更具挑战性，往往需要较多的工程经验与任务先验知识。在我们的工作中，我们遵循了 GIGAFLOW 的奖励设计方法 [2]，该设计在大规模实验中已被验证有效。

至于使用奖励模型（reward model）来学习人类偏好，我认为这在处理无法直接量化的人类行为准则时是值得进一步研究的方向。例如，学习人类在复杂交互中的潜在偏好，或对"舒适"的主观定义，可以通过奖励模型进行建模。然而，如果学习的目标本身就是可以明确量化的指标（如 collision、off-road 等），那么直接使用真实信号更为合理——在真值易得的前提下没有必要通过模型进行替代。只有在无法获取真实奖励信号的场景下，[3] 所提出的使用奖励模型来估计奖励才具有必要性和合理性。

[1] CaRL: Learning Scalable Planning Policies with Simple Rewards
[2] Robust Autonomy Emerges from Self-Play
[3] ReSim: Reliable World Simulation for Autonomous Driving
